File: lib/js/examples/encode.js
'use strict';

var Aeson_encode = require("../src/Aeson_encode.js");

console.log(JSON.stringify(/* array */[
          "foo",
          "bar"
        ]));

console.log(JSON.stringify(/* array */[
            "foo",
            "bar"
          ].map((function (prim) {
                return prim;
              }))));

console.log(Aeson_encode.object_(/* :: */[
          /* tuple */[
            "x",
            42
          ],
          /* :: */[
            /* tuple */[
              "foo",
              "bar"
            ],
            /* [] */0
          ]
        ]));

function point(r) {
  return Aeson_encode.object_(/* :: */[
              /* tuple */[
                "x",
                r[/* x */0]
              ],
              /* :: */[
                /* tuple */[
                  "y",
                  r[/* y */1]
                ],
                /* [] */0
              ]
            ]);
}

function line(r) {
  var match = r[/* thickness */2];
  return Aeson_encode.object_(/* :: */[
              /* tuple */[
                "start",
                point(r[/* start */0])
              ],
              /* :: */[
                /* tuple */[
                  "end",
                  point(r[/* end_ */1])
                ],
                /* :: */[
                  /* tuple */[
                    "thickness",
                    match !== undefined ? match : null
                  ],
                  /* [] */0
                ]
              ]
            ]);
}

var Encode = /* module */[
  /* point */point,
  /* line */line
];

var data = /* record */[
  /* start : record */[
    /* x */1.1,
    /* y */-0.4
  ],
  /* end_ : record */[
    /* x */5.3,
    /* y */3.8
  ],
  /* thickness */2
];

console.log(line(data));

exports.Encode = Encode;
exports.data = data;
/*  Not a pure module */

File: examples/encode.ml
(* Encoding a JSON data structure using Aeson.Encode *)

(* prints ["foo", "bar"] *)
let _ =
  [| "foo"; "bar" |]
  |> Aeson.Encode.stringArray
  |> Aeson.Json.stringify
  |> Js.log

(* prints ["foo", "bar"] *)
let _ =
  [| "foo"; "bar" |]
  |> Js.Array.map Aeson.Encode.string
  |> Aeson.Encode.array
  |> Aeson.Json.stringify
  |> Js.log

(* prints { x: 42, foo: 'bar' } *)
let _ =
  Aeson.Encode.(
    object_ [
      ("x", int 42);
      ("foo", string "bar")
    ]
    |> Js.log
  )

(* Advanced example: encode a record *)
type line = {
  start: point;
  end_: point;
  thickness: int option
}
and point = {
  x: float;
  y: float
}

module Encode = struct
  let point r =
    let open! Aeson.Encode in (
      object_ [
        ("x", float r.x);
        ("y", float r.y)
      ]
    )
  let line r =
    Aeson.Encode.(
      object_ [
        ("start", point r.start);
        ("end", point r.end_);
        ("thickness", match r.thickness with Some x -> int x | None -> null)
      ]
    )
end

let data = {
  start = { x = 1.1; y = -0.4 };
  end_ = { x = 5.3; y = 3.8 };
  thickness = Some 2
}

let _ =
  data
  |> Encode.line
  |> Js.log
